module/zstd/zstd.c

```diff
@@ -0,0 +1,610 @@
+	ZSTD_CCtx *cctx;
+
+	levelcookie = zstd_enum_to_cookie(n);
+	ASSERT3U(d_len, >=, sizeof (bufsiz));
+	ASSERT3U(d_len, <=, s_len);
+	ASSERT3U(levelcookie, !=, 0);
+
+	if (levelcookie == ZIO_COMPLEVEL_DEFAULT) {
+		levelcookie = ZIO_ZSTD_LEVEL_DEFAULT;
+	}
+	if (levelcookie == ZIO_ZSTDLVL_DEFAULT) {
+		levelcookie = ZIO_ZSTD_LEVEL_DEFAULT;
+	}
+
+	cctx = ZSTD_createCCtx_advanced(zstd_malloc);
+	/*
+	 * out of kernel memory, gently fall through - this will disable
+	 * compression in zio_compress_data
+	 */
+	if (cctx == NULL) {
+		return (s_len);
+	}
+
+	c_len = ZSTD_compressCCtx(cctx,
+	    &dest[sizeof (bufsiz) + sizeof (levelcookie)],
+	    d_len - sizeof (bufsiz) - sizeof (levelcookie),
+	    s_start, s_len, levelcookie);
+
+	ZSTD_freeCCtx(cctx);
+
+	/* Error in the compression routine. disable compression. */
+	if (ZSTD_isError(c_len)) {
+		return (s_len);
+	}
+
+	/*
+	 * Encode the compressed buffer size at the start. We'll need this in
+	 * decompression to counter the effects of padding which might be
+	 * added to the compressed buffer and which, if unhandled, would
+	 * confuse the hell out of our decompression function.
+	 */
+	bufsiz = c_len;
+	*(uint32_t *)dest = BE_32(bufsiz);
+	/*
+	 * Encode the compression level as well. We may need to know the
+	 * original compression level if compressed_arc is disabled, to match
+	 * the compression settings to write this block to the L2ARC.
+	 * Encode the actual level, so if the enum changes in the future,
+	 * we will be compatible.
+	 */
+	*(uint32_t *)(&dest[sizeof (bufsiz)]) = BE_32(levelcookie);
+
+	return (c_len + sizeof (bufsiz) + sizeof (levelcookie));
+}
+
+int
+zstd_get_level(void *s_start, size_t s_len, uint8_t *level)
+{
+	const char *src = s_start;
+	uint32_t levelcookie = BE_IN32(&src[sizeof (levelcookie)]);
+	uint8_t zstdlevel = zstd_cookie_to_enum(levelcookie);
+
+	ASSERT3U(zstdlevel, !=, ZIO_ZSTDLVL_INHERIT);
+
+	if (level != NULL) {
+		*level = zstdlevel;
+	}
+
+	return (0);
+}
+
+int
+zstd_decompress_level(void *s_start, void *d_start, size_t s_len, size_t d_len,
+    uint8_t *level)
+{
+	const char *src = s_start;
+	ZSTD_DCtx *dctx;
+	size_t result;
+	uint32_t bufsiz = BE_IN32(src);
+	int32_t levelcookie = (int32_t)BE_IN32(&src[sizeof (bufsiz)]);
+	uint8_t zstdlevel = zstd_cookie_to_enum(levelcookie);
+
+	ASSERT3U(d_len, >=, s_len);
+	ASSERT3U(zstdlevel, !=, ZIO_ZSTDLVL_INHERIT);
+
+	/* invalid compressed buffer size encoded at start */
+	if (bufsiz + sizeof (bufsiz) > s_len) {
+		return (1);
+	}
+
+	dctx = ZSTD_createDCtx_advanced(zstd_dctx_malloc);
+	if (dctx == NULL)
+		return (1);
+
+	result = ZSTD_decompressDCtx(dctx, d_start, d_len,
+	    &src[sizeof (bufsiz) + sizeof (levelcookie)], bufsiz);
+
+	ZSTD_freeDCtx(dctx);
+
+	/*
+	 * Returns 0 on success (decompression function returned non-negative)
+	 * and non-zero on failure (decompression function returned negative.
+	 */
+	if (ZSTD_isError(result)) {
+		return (1);
+	}
+
+	if (level != NULL) {
```

Comment:
did you mean `if (zstdlevel != NULL)`?

Reply:
no. level is a argument based return pointer. so if non NULL *level = blahblah

Reply:
ack